connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/inbound/DefaultProcessInstanceContext.java

```diff
@@ -0,0 +1,116 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.camunda.connector.runtime.core.inbound;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.camunda.connector.api.inbound.ProcessInstanceContext;
+import io.camunda.connector.api.validation.ValidationProvider;
+import io.camunda.connector.api.validation.ValidationUtil;
+import io.camunda.connector.feel.FeelEngineWrapper;
+import io.camunda.connector.feel.FeelEngineWrapperException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Supplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class DefaultProcessInstanceContext implements ProcessInstanceContext {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DefaultProcessInstanceContext.class);
+
+  private final Long processInstanceKey;
+  private final ValidationProvider validationProvider;
+  private final FeelEngineWrapper feelEngineWrapper;
+  private final ObjectMapper objectMapper;
+  private final Map<String, Object> contextProperties;
+  private final Supplier<Map<String, Object>> operatePropertiesSupplier;
+
+  public DefaultProcessInstanceContext(
+      final Long processInstanceKey,
+      final ValidationProvider validationProvider,
+      final FeelEngineWrapper feelEngineWrapper,
+      final ObjectMapper objectMapper,
+      final Map<String, Object> contextProperties,
+      final Supplier<Map<String, Object>> operateVariables) {
+    this.processInstanceKey = processInstanceKey;
+    this.validationProvider =
+        validationProvider == null
+            ? ValidationUtil.discoverDefaultValidationProviderImplementation()
+            : validationProvider;
+    this.feelEngineWrapper = feelEngineWrapper;
+    this.objectMapper = objectMapper;
+    this.contextProperties = contextProperties;
+    this.operatePropertiesSupplier = operateVariables;
+  }
+
+  @Override
+  public Long getKey() {
+    return processInstanceKey;
+  }
+
+  @Override
+  public <T> T bind(final Class<T> cls) {
+    HashMap<String, Object> copyOfProperties = new HashMap<>(contextProperties);
+    Map<String, Object> stringObjectMap = operatePropertiesSupplier.get();
+
+    copyOfProperties.putAll(stringObjectMap);
+    try {
+      if (evaluateAndPrepareForBinding(copyOfProperties)) {
+        T mappedObject = objectMapper.convertValue(copyOfProperties, cls);
+        validationProvider.validate(mappedObject);
+        return mappedObject;
+      } else {
+        LOGGER.debug(
+            "Evaluation and preparation for binding failed for processInstanceKey {}.",
+            processInstanceKey);
+        return null;
+      }
+    } catch (Exception e) {
+      LOGGER.warn(
+          "Error occurred while binding properties for processInstanceKey {}: {}",
+          processInstanceKey,
+          e.getMessage());
+      return null;
+    }
```

Comment:
You can remove the try and the `null` return value as this should just lead to an error.